### PR TITLE
openstackclient pod dynamic Ansible inventory

### PR DESCRIPTION
**Pull request by dprince**
This patch implements a dynamic Ansible inventory script
for the openstackclient pod. This inventory queries
the openstacknets API and parses out the relevant
host IPs which are then formatted in json for Ansible.

Docs are updated to recommend the new inventory is used
instead of having to rely on the static playbook generated
inventory we were using.

The motivation for this patch was to eliminate the dependency
on playbook generation to complete before node registration
could be accomplished, thus decoupling the two mechanism. This
dependency was a blocker for further deployment workflow changes.

### CODE REVIEW

**Comment by dprince** [comment]
This is very close but I seem to have broken something last minute as the pod hangs on init. Will have another look in the morning.

File: README.md [hunk 1]
```diff
@@ -518,7 +518,28 @@ Create a base RHEL data volume prior to deploying OpenStack.  This will be used
     oc create -f compute.yaml
     ```
 
-6) (optional) Create roles file
+6) Node registration (register the overcloud systems to required channels)
+
+    Wait for the above resource to finish deploying (Compute and ControlPlane). Once the resources finish deploying proceed with node registration.
+
+    Use the procedure as described in [5.9. Running Ansible-based registration manually](https://access.redhat.com/documentation/en-us/red_hat_openstack_platform/16.2/html-single/advanced_overcloud_customization/index#running-ansible-based-registration-manually-portal) do do so.
```
Comment (abays) [inline README.md:525]: s/do do so/to do so

File: README.md [hunk 2]
```diff
@@ -518,7 +518,28 @@ Create a base RHEL data volume prior to deploying OpenStack.  This will be used
...
+
+    ```bash
+    oc rsh openstackclient
+    bash
+    cd /home/cloud-admin
+
+    <create the ansible playbook for the overcloud nodes - e.g. rhsm.yaml>
+
+    # register the overcloud nodes to required repositories
+    ansible-playpook -i /home/cloud-admin/ctlplane-ansible-inventory ./rhsm.yaml
```
Comment (abays) [inline README.md:539]: s/playpook/playbook

File: templates/openstackclient/bin/init.sh [hunk 3]
```diff
@@ -44,6 +44,9 @@ if [ -d /mnt/ssh-config ]; then
   chown -R cloud-admin: /home/cloud-admin/.ssh
 fi
 
+# make the ctlplane ansible inventory executable
+chmod 755 /home/cloud-admin/ctlplane-ansible-inventory
```
Comment (olliewalsh) [inline templates/openstackclient/bin/init.sh:48]: Is this necessary? The ConfigMapVolumeSource default mode is 0755.

File: README.md (hunk 2 above)
Comment (dprince) [reply README.md:539]: But pook is so much better! Just kidding, will fix it

File: README.md (hunk 1 above)
Comment (dprince) [reply README.md:525]: got it

File: templates/openstackclient/bin/ansible-inventory [hunk 4]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
+
+APISERVER=https://$KUBERNETES_SERVICE_HOST
+SERVICEACCOUNT=/var/run/secrets/kubernetes.io/serviceaccount
+NAMESPACE=$(cat ${SERVICEACCOUNT}/namespace)
+TOKEN=$(cat ${SERVICEACCOUNT}/token)
+CACERT=${SERVICEACCOUNT}/ca.crt
+
+HOSTS=$(curl -s --cacert ${CACERT} --header "Authorization: Bearer ${TOKEN}" -X GET ${APISERVER}/apis/osp-director.openstack.org/v1beta1/namespaces/$NAMESPACE/openstacknets/ctlplane | jq .spec.roleReservations[].reservations[] | jq 'del(. | select(.hostname == "$OSP_DIRECTOR_OPERATOR_IPSET_HOSTNAME"))' | jq -er 'select(. | . != null).ip' | xargs)
```
[the remark this answers is not in the document]
Comment (dprince) [reply templates/openstackclient/bin/ansible-inventory]: @stuggi have a look at the latest version. Should now support the individual roles while still supporting 'overcloud' (all roles). And thanks for the jq query updates. I merged those in.

File: templates/openstackclient/bin/init.sh (hunk 3 above)
Comment (dprince) [reply templates/openstackclient/bin/init.sh]: Good call. This was actually what broke it so removing it fixes everything.

**Comment by dprince** [comment]
added the roles feature, and then ansible vars for ssh. still no dice on this yet. will have a closer look at tripleo-ansible-inventory.yaml tomorrow and see what might be missing

File: templates/openstackclient/bin/ansible-inventory [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+for ROLE in $ROLES; do
+ROLE_IPS=$(echo $ROLE_RESERVATIONS | jq -r "select(.key == \"$ROLE\") | .value.reservations[].ip")
+cat<<EOF
+  "$ROLE": {
+        "hosts": [
+EOF
+
+    printHostIPs "$ROLE" "$ROLE_IPS"
+
+echo '         ]'
```
Comment (stuggi) [inline templates/openstackclient/bin/ansible-inventory:59]: needs to be `echo '         ],'` as vars follows

File: controllers/openstackclient_controller.go [hunk 6]
```diff
@@ -414,6 +414,8 @@ func (r *OpenStackClientReconciler) podCreateOrUpdate(
 
 	(*envVars)["KOLLA_CONFIG_STRATEGY"] = common.EnvValue("COPY_ALWAYS")
 
+	(*envVars)["OSP_DIRECTOR_OPERATOR_IPSET_HOSTNAME"] = common.EnvValue(hostname)
+
```
Comment (stuggi) [inline controllers/openstackclient_controller.go:418]: do we still need this?

Comment (dprince) [reply controllers/openstackclient_controller.go:418]: no, I meant to drop it.

File: templates/openstackclient/bin/ansible-inventory (hunk 5 above)
Comment (olliewalsh) [reply templates/openstackclient/bin/ansible-inventory:59]: is it worth rewriting this in python or $(something that can serialise to json)?

Comment (dprince) [reply templates/openstackclient/bin/ansible-inventory:59]: ack. I had already fixed this locally and just not pushed it yet. Still this functionally isn't working quite yet.

**Comment by dprince** [comment]
This is a go. Seems to work fine now that I reinstalled OCP.

File: templates/openstackclient/bin/ansible-inventory [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+if [ "$1" == "--list" ] ; then
+cat<<EOF
+{
+  "overcloud": {
+        "children": [
+EOF
+
+
+# print out the 'ALL' section (contains each rolename)
```
Comment (dprince) [inline templates/openstackclient/bin/ansible-inventory:37]: this should say 'overcloud' not 'ALL'

**Review by stuggi — APPROVED** [review]
/lgtm

**Pull request merged**